drivers/platform/xilinx/i2c.c

```diff
@@ -57,65 +67,98 @@
  * @return SUCCESS in case of success, FAILURE otherwise.
  */
 int32_t i2c_init(struct i2c_desc **desc,
-		 const struct i2c_init_param *param)
+		 const struct i2c_init_param *init_param)
 {
+	int32_t		ret;
+	i2c_desc	*idesc;
+	xil_i2c_desc	*xdesc;
+	xil_i2c_init	*xinit;
+
+	idesc = (struct i2c_desc *)malloc(sizeof(*idesc));
+	xdesc = (struct xil_i2c_desc *)malloc(sizeof(*xdesc));
+
+	if(!idesc || !xdesc)
+		goto error;
+
+	idesc->max_speed_hz = init_param->max_speed_hz;
+	idesc->slave_address = init_param->slave_address;
+	xinit = init_param->extra;
+	idesc->extra = xdesc;
+
+	switch (xinit->type) {
+	case IIC_PL:
 #ifdef XIIC_H
-	i2c_desc *dev;
-	xil_i2c_desc *xil_dev;
-	xil_i2c_init_param *xil_param;
-	int32_t ret;
+		xdesc->instance = (XIic *)malloc(sizeof(XIic));
+		if(!xdesc->instance)
+			goto pl_error;
 
-	dev = calloc(1, sizeof *dev);
-	if(!dev)
-		return FAILURE;
+		xdesc->config = XIic_LookupConfig(xinit->device_id);
+		if(xdesc->config == NULL)
+			goto pl_error;
 
-	dev->extra = calloc(1, sizeof *xil_dev);
-	if(!(dev->extra)) {
-		free(dev);
-		return FAILURE;
-	}
+		ret = XIic_CfgInitialize(xdesc->instance,
+					 xdesc->config,
+					 ((XIic_Config*)xdesc->config)
+					 ->BaseAddress);
+		if(ret != SUCCESS)
+			goto pl_error;
 
-	xil_dev = dev->extra;
-	xil_param = param->extra;
+		Xil_ExceptionInit();
+		Xil_ExceptionRegisterHandler(XIL_EXCEPTION_ID_IRQ_INT,
+					     XIic_InterruptHandler,
+					     &xdesc->instance);
```

Comment:
Also the `instance` member of `xdesc` is already a pointer (void pointer) to the struct we need. Giving it by reference here would make it a double pointer. Reference needs to be removed and maybe cast is needed.